File: frontend/js/views/lean_fire.js
// Lean FIRE — minimalist retirement variant (≤ $40k annual expenses).
// Validates the expense plan is within lean tier and reports FI
// number + years-to-target.

import { api } from '../api.js';
import { esc } from '../util.js';
import { t } from '../i18n.js';

export async function renderLeanFire(mount, _state) {
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.lean_fire.title">// LEAN FIRE</span></h1>
        <p class="muted small" data-i18n-html="view.lean_fire.intro">
            <strong>Lean FIRE</strong> = FIRE at minimalist spending — typically annual
            expenses ≤ <strong>$40k</strong>. Smaller portfolio target, faster timeline,
            less margin if expenses surprise. Tiers: <strong>ultralean</strong> ≤ $25k,
            <strong>lean</strong> ≤ $40k, <strong>borderline</strong> ≤ $55k, otherwise
            <strong>not lean</strong>.
        </p>
        <div class="chart-panel">
            <div style="display:grid;grid-template-columns:repeat(auto-fit,minmax(200px,1fr));gap:8px;margin-bottom:12px">
                <label><span class="muted small" data-i18n="view.lean_fire.field.nw">Current net worth $</span>
                    <input type="number" id="lf-nw" step="1000" min="0" value="200000" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.lean_fire.field.expenses">Annual expenses $</span>
                    <input type="number" id="lf-expenses" step="1000" min="0" value="30000" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.lean_fire.field.swr">SWR %</span>
                    <input type="number" id="lf-swr" step="0.25" min="0.25" max="20" value="4" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.lean_fire.field.return">Real return %/yr</span>
                    <input type="number" id="lf-return" step="0.25" min="-10" max="20" value="5" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.lean_fire.field.contrib">Monthly contribution $</span>
                    <input type="number" id="lf-contrib" step="50" min="0" value="1500" style="width:100%"></label>
            </div>
            <button class="btn btn-sm primary" id="lf-run" data-shortcut="r" data-i18n="view.lean_fire.btn.run">⚡ Compute Lean FI</button>
            <div id="lf-result"></div>
        </div>
    `;
    mount.querySelector('#lf-run').addEventListener('click', () => runCompute(mount));
    await runCompute(mount);
}

async function runCompute(mount) {
    const result = mount.querySelector('#lf-result');
    const input = {
        current_nw_usd: parseFloat(mount.querySelector('#lf-nw').value) || 0,
        annual_expenses_usd: parseFloat(mount.querySelector('#lf-expenses').value) || 0,
        safe_withdrawal_rate_pct: parseFloat(mount.querySelector('#lf-swr').value) || 4,
        expected_real_return_pct: parseFloat(mount.querySelector('#lf-return').value) || 5,
        monthly_contribution_usd: parseFloat(mount.querySelector('#lf-contrib').value) || 0,
    };
    result.innerHTML = `<p class="muted">${esc(t('view.lean_fire.status.computing'))}</p>`;
    try {
        const r = await api('/lean-fire/compute', { method: 'POST', body: JSON.stringify(input) });
        const fiCls = r.is_lean_fi ? 'pos' : 'neg';
        const tierCls = r.expense_tier === 'ultralean' || r.expense_tier === 'lean' ? 'pos'
                       : r.expense_tier === 'not_lean' ? 'neg' : '';
        const deltaCls = r.current_vs_fi_delta_usd >= 0 ? 'pos' : 'neg';
        result.innerHTML = `
            <div style="display:grid;grid-template-columns:repeat(auto-fit,minmax(200px,1fr));gap:12px;margin-top:1rem">
                <div><div class="muted small">${esc(t('view.lean_fire.field.fi_number'))}</div>
                    <strong style="font-size:1.4em">$${(r.fi_number_usd / 1000).toFixed(0)}K</strong></div>
                <div><div class="muted small">${esc(t('view.lean_fire.field.delta'))}</div>
                    <strong class="${deltaCls}">${r.current_vs_fi_delta_usd >= 0 ? '+' : '−'}$${(Math.abs(r.current_vs_fi_delta_usd) / 1000).toFixed(1)}K</strong></div>
                <div><div class="muted small">${esc(t('view.lean_fire.field.is_fi'))}</div>
                    <strong class="${fiCls}">${r.is_lean_fi ? '✓ ' + esc(t('view.lean_fire.status.lean_fi')) : '✗ ' + esc(t('view.lean_fire.status.not_yet'))}</strong></div>
                <div><div class="muted small">${esc(t('view.lean_fire.field.years_to'))}</div>
                    <strong>${r.years_to_fi == null ? '∞' : r.years_to_fi.toFixed(1) + 'y'}</strong></div>
                <div><div class="muted small">${esc(t('view.lean_fire.field.tier'))}</div>
                    <strong class="${tierCls}" style="text-transform:uppercase">${esc(t('view.lean_fire.tier.' + r.expense_tier) || r.expense_tier)}</strong></div>
                <div><div class="muted small">${esc(t('view.lean_fire.field.cap'))}</div>
                    <strong>$${r.upper_lean_threshold_usd.toFixed(0)}/yr</strong></div>
            </div>
        `;
    } catch (e) {
        result.innerHTML = `<p class="neg">${esc(String(e))}</p>`;
    }
}

File: crates/traderview-db/src/lean_fire.rs
//! Lean FIRE calculator.
//!
//! Lean FIRE is the FIRE variant aimed at minimalist retirement —
//! typically annual expenses ≤ $40k (loose community-published
//! threshold; some define < $25k or < median household expenses).
//! Low spending → smaller portfolio target → faster timeline → but
//! less margin for error if expenses surprise.
//!
//! This module is purposely focused on the minimalist case: validates
//! the user's expense plan is within the lean tier, computes the FI
//! number + years-to-target + classifies the user's expense level as
//! ultralean / lean / borderline / not_lean.
//!
//! Pure compute.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct LeanFireInput {
    pub current_nw_usd: f64,
    pub annual_expenses_usd: f64,
    #[serde(default = "default_swr")]
    pub safe_withdrawal_rate_pct: f64,
    #[serde(default = "default_return")]
    pub expected_real_return_pct: f64,
    #[serde(default)]
    pub monthly_contribution_usd: f64,
}

fn default_swr() -> f64 { 4.0 }
fn default_return() -> f64 { 5.0 }

#[derive(Debug, Clone, Serialize)]
pub struct LeanFireReport {
    pub fi_number_usd: f64,
    pub current_vs_fi_delta_usd: f64,
    pub is_lean_fi: bool,
    pub years_to_fi: Option<f64>,
    pub expense_tier: &'static str,
    pub upper_lean_threshold_usd: f64,
}

// ─── Pure compute ─────────────────────────────────────────────────────────

const ULTRALEAN_CAP: f64 = 25_000.0;
const LEAN_CAP: f64 = 40_000.0;
const BORDERLINE_CAP: f64 = 55_000.0;

pub fn fi_number(annual_expenses: f64, swr_pct: f64) -> f64 {
    if swr_pct <= 0.0 { return 0.0; }
    annual_expenses / (swr_pct / 100.0)
}

pub fn classify_expense_tier(annual_expenses: f64) -> &'static str {
    if annual_expenses <= ULTRALEAN_CAP { "ultralean" }
    else if annual_expenses <= LEAN_CAP { "lean" }
    else if annual_expenses <= BORDERLINE_CAP { "borderline" }
    else { "not_lean" }
}

pub fn years_until_target(
    current: f64,
    monthly_contribution: f64,
    annual_return_pct: f64,
    target: f64,
) -> Option<f64> {
    if current >= target { return Some(0.0); }
    if monthly_contribution <= 0.0 && annual_return_pct <= 0.0 { return None; }
    let r = annual_return_pct / 100.0;
    let monthly_r = r / 12.0;
    let mut years = 0.0_f64;
    let step = 1.0 / 12.0;
    while years < 100.0 {
        years += step;
        let months = years * 12.0;
        let lump = current * (1.0 + r).powf(years);
        let annuity = if monthly_r.abs() < 1e-12 {
            monthly_contribution * months
        } else {
            monthly_contribution * ((1.0 + monthly_r).powf(months) - 1.0) / monthly_r
        };
        if lump + annuity >= target { return Some(years); }
    }
    None
}

pub fn compute(input: &LeanFireInput) -> LeanFireReport {
    let fi = fi_number(input.annual_expenses_usd, input.safe_withdrawal_rate_pct);
    let delta = input.current_nw_usd - fi;
    let is_fi = input.current_nw_usd >= fi - 0.005 && fi > 0.0;
    let years = years_until_target(
        input.current_nw_usd,
        input.monthly_contribution_usd,
        input.expected_real_return_pct,
        fi,
    );
    let tier = classify_expense_tier(input.annual_expenses_usd);
    LeanFireReport {
        fi_number_usd: fi,
        current_vs_fi_delta_usd: delta,
        is_lean_fi: is_fi,
        years_to_fi: years,
        expense_tier: tier,
        upper_lean_threshold_usd: LEAN_CAP,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn input() -> LeanFireInput {
        LeanFireInput {
            current_nw_usd: 200_000.0,
            annual_expenses_usd: 30_000.0,
            safe_withdrawal_rate_pct: 4.0,
            expected_real_return_pct: 5.0,
            monthly_contribution_usd: 1_500.0,
        }
    }

    #[test]
    fn fi_number_basic() {
        assert_eq!(fi_number(30_000.0, 4.0), 750_000.0);
    }

    #[test]
    fn classify_expense_tier_thresholds() {
        assert_eq!(classify_expense_tier(20_000.0), "ultralean");
        assert_eq!(classify_expense_tier(25_000.0), "ultralean");
        assert_eq!(classify_expense_tier(35_000.0), "lean");
        assert_eq!(classify_expense_tier(40_000.0), "lean");
        assert_eq!(classify_expense_tier(50_000.0), "borderline");
        assert_eq!(classify_expense_tier(80_000.0), "not_lean");
    }

    #[test]
    fn years_until_target_already_there() {
        assert_eq!(years_until_target(500_000.0, 0.0, 5.0, 100_000.0), Some(0.0));
    }

    #[test]
    fn years_until_target_impossible() {
        assert!(years_until_target(100.0, 0.0, 0.0, 1_000_000.0).is_none());
    }

    #[test]
    fn compute_lean_30k_fi() {
        let r = compute(&input());
        assert_eq!(r.fi_number_usd, 750_000.0);
        assert_eq!(r.expense_tier, "lean");
        assert_eq!(r.upper_lean_threshold_usd, 40_000.0);
    }

    #[test]
    fn compute_not_yet_at_200k() {
        let r = compute(&input());
        assert!(!r.is_lean_fi);
        assert!(r.current_vs_fi_delta_usd < 0.0);
    }

    #[test]
    fn compute_already_fi_at_800k() {
        let mut i = input();
        i.current_nw_usd = 800_000.0;
        let r = compute(&i);
        assert!(r.is_lean_fi);
        assert_eq!(r.years_to_fi, Some(0.0));
    }

    #[test]
    fn compute_ultralean_classification() {
        let mut i = input();
        i.annual_expenses_usd = 20_000.0;
        let r = compute(&i);
        assert_eq!(r.expense_tier, "ultralean");
    }

    #[test]
    fn compute_not_lean_classification() {
        let mut i = input();
        i.annual_expenses_usd = 80_000.0;
        let r = compute(&i);
        assert_eq!(r.expense_tier, "not_lean");
    }

    #[test]
    fn compute_years_to_fi_positive_for_realistic() {
        let r = compute(&input());
        assert!(r.years_to_fi.is_some());
        let y = r.years_to_fi.unwrap();
        assert!(y > 5.0 && y < 30.0);
    }
}
